File: processing/image_editor.py
import os
import logging
from PIL import Image, ImageEnhance, ImageFilter
import rawpy
import numpy as np

logger = logging.getLogger(__name__)

class ImageEditor:
    """
    Class for handling image editing operations on both RAW and regular image files.
    """
    
    def __init__(self, input_path):
        """
        Initialize the image editor with an input file.
        
        Args:
            input_path (str): Path to the input image file
        """
        self.input_path = input_path
        self.image = None
        self.original_image = None
        self.is_raw = self._check_if_raw()
        self.metadata = {}
        self.edit_history = []
        
        # Load the image
        self.load_image()
    
    def _check_if_raw(self):
        """Check if the input file is a RAW format based on extension."""
        raw_extensions = ['.arw', '.cr2', '.crw', '.dng', '.nef', '.orf', '.pef', '.raf', '.raw', '.rw2', '.srw']
        ext = os.path.splitext(self.input_path)[1].lower()
        return ext in raw_extensions
    
    def load_image(self):
        """Load the image from the input path."""
        try:
            if self.is_raw:
                # Use rawpy to load RAW file
                with rawpy.imread(self.input_path) as raw:
                    rgb = raw.postprocess(
                        use_camera_wb=True,
                        no_auto_bright=False,
                        output_bps=8
                    )
                    self.image = Image.fromarray(rgb)
                    self.original_image = self.image.copy()
                    self.metadata['raw_sizes'] = raw.sizes
                    self.metadata['raw_color_desc'] = raw.color_desc
            else:
                # Use PIL to load regular image
                self.image = Image.open(self.input_path)
                self.original_image = self.image.copy()
                
            # Store basic metadata
            self.metadata['format'] = self.image.format
            self.metadata['mode'] = self.image.mode
            self.metadata['size'] = self.image.size
            
            logger.info(f"Successfully loaded image: {self.input_path}")
            return True
        except Exception as e:
            logger.error(f"Error loading image {self.input_path}: {e}", exc_info=True)
            return False
    
    def get_preview(self, max_size=800):
        """
        Get a resized preview of the current image state.
        
        Args:
            max_size (int): Maximum dimension (width or height) for the preview
            
        Returns:
            PIL.Image: Resized preview image
        """
        if not self.image:
            return None
            
        # Calculate new dimensions while maintaining aspect ratio
        width, height = self.image.size
        if width > height:
            new_width = min(width, max_size)
            new_height = int(height * (new_width / width))
        else:
            new_height = min(height, max_size)
            new_width = int(width * (new_height / height))
            
        # Return a resized copy
        return self.image.resize((new_width, new_height), Image.LANCZOS)
    
    def crop(self, left, top, right, bottom):
        """
        Crop the current image.
        
        Args:
            left, top, right, bottom (int): Crop coordinates
            
        Returns:
            bool: Success status
        """
        try:
            if not self.image:
                return False
                
            # Store the action for history
            self.edit_history.append(('crop', (left, top, right, bottom)))
            
            # Perform the crop
            self.image = self.image.crop((left, top, right, bottom))
            return True
        except Exception as e:
            logger.error(f"Error cropping image: {e}", exc_info=True)
            return False
    
    def adjust_brightness(self, factor):
        """
        Adjust image brightness.
        
        Args:
            factor (float): Brightness factor (0.0-2.0, 1.0 is original)
            
        Returns:
            bool: Success status
        """
        try:
            if not self.image:
                return False
                
            # Store the action for history
            self.edit_history.append(('brightness', factor))
            
            # Apply brightness adjustment
            enhancer = ImageEnhance.Brightness(self.image)
            self.image = enhancer.enhance(factor)
            return True
        except Exception as e:
            logger.error(f"Error adjusting brightness: {e}", exc_info=True)
            return False
    
    def adjust_contrast(self, factor):
        """
        Adjust image contrast.
        
        Args:
            factor (float): Contrast factor (0.0-2.0, 1.0 is original)
            
        Returns:
            bool: Success status
        """
        try:
            if not self.image:
                return False
                
            # Store the action for history
            self.edit_history.append(('contrast', factor))
            
            # Apply contrast adjustment
            enhancer = ImageEnhance.Contrast(self.image)
            self.image = enhancer.enhance(factor)
            return True
        except Exception as e:
            logger.error(f"Error adjusting contrast: {e}", exc_info=True)
            return False
    
    def adjust_saturation(self, factor):
        """
        Adjust image saturation.
        
        Args:
            factor (float): Saturation factor (0.0-2.0, 1.0 is original)
            
        Returns:
            bool: Success status
        """
        try:
            if not self.image:
                return False
                
            # Store the action for history
            self.edit_history.append(('saturation', factor))
            
            # Apply saturation adjustment
            enhancer = ImageEnhance.Color(self.image)
            self.image = enhancer.enhance(factor)
            return True
        except Exception as e:
            logger.error(f"Error adjusting saturation: {e}", exc_info=True)
            return False
    
    def rotate(self, angle, expand=True):
        """
        Rotate the image.
        
        Args:
            angle (float): Rotation angle in degrees
            expand (bool): Whether to expand the output to fit the rotated image
            
        Returns:
            bool: Success status
        """
        try:
            if not self.image:
                return False
                
            # Store the action for history
            self.edit_history.append(('rotate', (angle, expand)))
            
            # Apply rotation
            self.image = self.image.rotate(angle, expand=expand, resample=Image.BICUBIC)
            return True
        except Exception as e:
            logger.error(f"Error rotating image: {e}", exc_info=True)
            return False
    
    def resize(self, width, height, maintain_aspect=True):
        """
        Resize the image.
        
        Args:
            width (int): Target width
            height (int): Target height
            maintain_aspect (bool): Whether to maintain aspect ratio
            
        Returns:
            bool: Success status
        """
        try:
            if not self.image:
                return False
                
            # Calculate new dimensions if maintaining aspect ratio
            if maintain_aspect:
                orig_width, orig_height = self.image.size
                ratio = min(width / orig_width, height / orig_height)
                new_width = int(orig_width * ratio)
                new_height = int(orig_height * ratio)
            else:
                new_width, new_height = width, height
                
            # Store the action for history
            self.edit_history.append(('resize', (new_width, new_height)))
            
            # Apply resize
            self.image = self.image.resize((new_width, new_height), Image.LANCZOS)
            return True
        except Exception as e:
            logger.error(f"Error resizing image: {e}", exc_info=True)
            return False
    
    def apply_filter(self, filter_type):
        """
        Apply a filter to the image.
        
        Args:
            filter_type (str): Filter type ('blur', 'sharpen', 'contour', etc.)
            
        Returns:
            bool: Success status
        """
        try:
            if not self.image:
                return False
                
            # Store the action for history
            self.edit_history.append(('filter', filter_type))
            
            # Apply the appropriate filter
            if filter_type == 'blur':
                self.image = self.image.filter(ImageFilter.BLUR)
            elif filter_type == 'sharpen':
                self.image = self.image.filter(ImageFilter.SHARPEN)
            elif filter_type == 'contour':
                self.image = self.image.filter(ImageFilter.CONTOUR)
            elif filter_type == 'edge_enhance':
                self.image = self.image.filter(ImageFilter.EDGE_ENHANCE)
            elif filter_type == 'emboss':
                self.image = self.image.filter(ImageFilter.EMBOSS)
            elif filter_type == 'smooth':
                self.image = self.image.filter(ImageFilter.SMOOTH)
            elif filter_type == 'detail':
                self.image = self.image.filter(ImageFilter.DETAIL)
            else:
                logger.warning(f"Unknown filter type: {filter_type}")
                return False
                
            return True
        except Exception as e:
            logger.error(f"Error applying filter {filter_type}: {e}", exc_info=True)
            return False
    
    def reset(self):
        """
        Reset to the original image state.
        
        Returns:
            bool: Success status
        """
        try:
            if not self.original_image:
                return False
                
            # Clear edit history
            self.edit_history = []
            
            # Reset to original
            self.image = self.original_image.copy()
            return True
        except Exception as e:
            logger.error(f"Error resetting image: {e}", exc_info=True)
            return False
    
    def undo(self):
        """
        Undo the last edit operation.
        
        Returns:
            bool: Success status
        """
        try:
            if not self.edit_history:
                return False
                
            # Reset and replay all actions except the last one
            self.image = self.original_image.copy()
            history = self.edit_history[:-1]
            self.edit_history = []
            
            # Replay all actions
            for action, params in history:
                if action == 'crop':
                    self.crop(*params)
                elif action == 'brightness':
                    self.adjust_brightness(params)
                elif action == 'contrast':
                    self.adjust_contrast(params)
                elif action == 'saturation':
                    self.adjust_saturation(params)
                elif action == 'rotate':
                    self.rotate(*params)
                elif action == 'resize':
                    self.resize(*params)
                elif action == 'filter':
                    self.apply_filter(params)
            
            return True
        except Exception as e:
            logger.error(f"Error undoing last action: {e}", exc_info=True)
            return False
    
    def save(self, output_path, format=None, quality=95):
        """
        Save the edited image.
        
        Args:
            output_path (str): Path to save the output image
            format (str): Output format (JPEG, PNG, etc.)
            quality (int): JPEG quality (1-100)
            
        Returns:
            bool: Success status
        """
        try:
            if not self.image:
                return False
                
            # Create output directory if it doesn't exist
            os.makedirs(os.path.dirname(os.path.abspath(output_path)), exist_ok=True)
            
            # Determine format from output path if not specified
            if not format:
                format = os.path.splitext(output_path)[1][1:].upper()
                if not format:
                    format = 'JPEG'
            
            # Save options
            save_opts = {}
            if format.upper() == 'JPEG':
                save_opts['quality'] = quality
            
            # Save the image
            self.image.save(output_path, format=format, **save_opts)
            logger.info(f"Successfully saved image to {output_path}")
            return True
        except Exception as e:
            logger.error(f"Error saving image to {output_path}: {e}", exc_info=True)
            return False

File: app/static/js/editor/image_editor.js
/**
 * Image Editor JavaScript Module
 * Handles UI interactions for image editing functionality
 */

// --- DOM Elements ---
// Image browser elements
const imageBrowser = document.getElementById('image-browser');
const btnRefreshImages = document.getElementById('btn-refresh-images');
const btnParentDir = document.getElementById('btn-parent-dir');
const currentPath = document.getElementById('current-path');
const directoryList = document.getElementById('directory-list');
const imageList = document.getElementById('image-list');

// Image preview elements
const previewContainer = document.getElementById('preview-container');
const imagePreview = document.getElementById('image-preview');
const previewPlaceholder = document.getElementById('preview-placeholder');
const previewLoading = document.getElementById('preview-loading');
const imageName = document.getElementById('image-name');

// Control elements
const btnUndo = document.getElementById('btn-undo');
const btnReset = document.getElementById('btn-reset');
const btnToggleCrop = document.getElementById('btn-toggle-crop');
const btnRotate = document.getElementById('btn-rotate');
const btnSave = document.getElementById('btn-save');

// Crop controls
const cropControls = document.getElementById('crop-controls');
const cropLeft = document.getElementById('crop-left');
const cropTop = document.getElementById('crop-top');
const cropRight = document.getElementById('crop-right');
const cropBottom = document.getElementById('crop-bottom');
const btnCropPreview = document.getElementById('btn-crop-preview');
const btnApplyCrop = document.getElementById('btn-apply-crop');

// Adjustment controls
const brightnessSlider = document.getElementById('brightness-slider');
const brightnessValue = document.getElementById('brightness-value');
const contrastSlider = document.getElementById('contrast-slider');
const contrastValue = document.getElementById('contrast-value');
const saturationSlider = document.getElementById('saturation-slider');
const saturationValue = document.getElementById('saturation-value');
const filterSelect = document.getElementById('filter-select');
const formatSelect = document.getElementById('format-select');
const qualitySlider = document.getElementById('quality-slider');
const qualityValue = document.getElementById('quality-value');
const outputFilename = document.getElementById('output-filename');
const btnShowMetadata = document.getElementById('btn-show-metadata');

// Modal elements
const metadataModal = document.getElementById('metadata-modal');
const closeMetadataModal = document.getElementById('close-metadata-modal');
const metadataContent = document.getElementById('metadata-content');
const rotateModal = document.getElementById('rotate-modal');
const closeRotateModal = document.getElementById('close-rotate-modal');
const rotationSlider = document.getElementById('rotation-slider');
const rotationValue = document.getElementById('rotation-value');
const btnApplyRotation = document.getElementById('btn-apply-rotation');

// --- State ---
let currentDirectory = '';
let selectedImage = null;
let imageMetadata = null;
let sessionId = 'image-editor-' + Date.now(); // Generate a unique session ID
let originalImageSize = null; // Store original image dimensions
let isCropMode = false;
let cropPreviewMode = false;

// --- API Functions ---

/**
 * List files and directories at the given path
 */
async function browseDirectory(directory = '') {
    try {
        showDirectoryLoading(true);
        
        const url = `/api/editor/preview/browse?dir=${encodeURIComponent(directory)}`;
        const response = await fetch(url);
        
        if (!response.ok) {
            throw new Error(`Error listing directory: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            currentDirectory = data.contents.path;
            updateBrowserUI(data.contents);
        } else {
            throw new Error(data.message || 'Failed to browse directory');
        }
    } catch (error) {
        console.error('Browse error:', error);
        alert(`Error browsing directory: ${error.message}`);
    } finally {
        showDirectoryLoading(false);
    }
}

/**
 * Load an image for editing
 */
async function loadImage(imagePath) {
    try {
        showImageLoading(true);
        
        // Close previous session if exists
        if (selectedImage) {
            try {
                await fetch('/api/editor/image/close', {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({ session_id: sessionId })
                });
            } catch (e) {
                console.warn('Error closing previous session:', e);
            }
        }
        
        // Create a new editing session
        const response = await fetch('/api/editor/image/load', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                path: imagePath,
                session_id: sessionId
            })
        });
        
        if (!response.ok) {
            throw new Error(`Error loading image: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            selectedImage = imagePath;
            imageMetadata = data.metadata;
            originalImageSize = imageMetadata.size;
            
            // Update UI
            updateImagePreview();
            updateImageInfo();
            enableControls(true);
        } else {
            throw new Error(data.message || 'Failed to load image');
        }
    } catch (error) {
        console.error('Load error:', error);
        alert(`Error loading image: ${error.message}`);
        showImageLoading(false);
    }
}

/**
 * Update the image preview
 */
async function updateImagePreview() {
    if (!selectedImage) {
        return;
    }
    
    try {
        showImageLoading(true);
        
        // Clear any existing preview
        imagePreview.src = '';
        
        // Calculate preview size to fit container
        const containerWidth = previewContainer.clientWidth;
        const maxSize = Math.min(800, containerWidth);
        
        // Set the preview image source with the current timestamp to avoid caching
        const timestamp = Date.now();
        imagePreview.src = `/api/editor/image/preview?session_id=${sessionId}&max_size=${maxSize}&t=${timestamp}`;
        
        // Show the preview when loaded
        imagePreview.onload = function() {
            showImageLoading(false);
            imagePreview.classList.remove('hidden');
            previewPlaceholder.classList.add('hidden');
        };
        
        // Handle load error
        imagePreview.onerror = function() {
            showImageLoading(false);
            imagePreview.classList.add('hidden');
            previewPlaceholder.classList.remove('hidden');
            previewPlaceholder.textContent = 'Error loading preview';
        };
    } catch (error) {
        console.error('Preview error:', error);
        showImageLoading(false);
        imagePreview.classList.add('hidden');
        previewPlaceholder.classList.remove('hidden');
        previewPlaceholder.textContent = 'Error loading preview';
    }
}

/**
 * Apply crop to the image
 */
async function applyCrop() {
    if (!selectedImage || !isCropMode) {
        return;
    }
    
    try {
        showImageLoading(true);
        
        // Get crop coordinates
        const left = parseInt(cropLeft.value) || 0;
        const top = parseInt(cropTop.value) || 0;
        const right = parseInt(cropRight.value) || 0;
        const bottom = parseInt(cropBottom.value) || 0;
        
        // Validate crop coordinates
        if (left >= right || top >= bottom) {
            throw new Error('Invalid crop coordinates. Right/Bottom must be greater than Left/Top.');
        }
        
        const response = await fetch('/api/editor/image/crop', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                session_id: sessionId,
                crop: [left, top, right, bottom]
            })
        });
        
        if (!response.ok) {
            throw new Error(`Error applying crop: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            // Update preview and exit crop mode
            updateImagePreview();
            toggleCropMode(false);
        } else {
            throw new Error(data.message || 'Failed to apply crop');
        }
    } catch (error) {
        console.error('Crop error:', error);
        alert(`Error applying crop: ${error.message}`);
        showImageLoading(false);
    }
}

/**
 * Apply rotation to the image
 */
async function applyRotation(angle) {
    if (!selectedImage) {
        return;
    }
    
    try {
        showImageLoading(true);
        
        const response = await fetch('/api/editor/image/rotate', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                session_id: sessionId,
                angle: angle,
                expand: true
            })
        });
        
        if (!response.ok) {
            throw new Error(`Error applying rotation: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            // Update preview
            updateImagePreview();
        } else {
            throw new Error(data.message || 'Failed to apply rotation');
        }
    } catch (error) {
        console.error('Rotation error:', error);
        alert(`Error applying rotation: ${error.message}`);
        showImageLoading(false);
    }
}

/**
 * Apply image adjustments
 */
async function applyAdjustments() {
    if (!selectedImage) {
        return;
    }
    
    try {
        showImageLoading(true);
        
        const adjustments = {
            brightness: parseFloat(brightnessSlider.value),
            contrast: parseFloat(contrastSlider.value),
            saturation: parseFloat(saturationSlider.value)
        };
        
        const response = await fetch('/api/editor/image/adjust', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                session_id: sessionId,
                adjustments: adjustments
            })
        });
        
        if (!response.ok) {
            throw new Error(`Error applying adjustments: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            // Update preview
            updateImagePreview();
        } else {
            throw new Error(data.message || 'Failed to apply adjustments');
        }
    } catch (error) {
        console.error('Adjustment error:', error);
        alert(`Error applying adjustments: ${error.message}`);
        showImageLoading(false);
    }
}

/**
 * Apply a filter to the image
 */
async function applyFilter(filterType) {
    if (!selectedImage || !filterType) {
        return;
    }
    
    try {
        showImageLoading(true);
        
        const response = await fetch('/api/editor/image/filter', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                session_id: sessionId,
                filter: filterType
            })
        });
        
        if (!response.ok) {
            throw new Error(`Error applying filter: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            // Update preview
            updateImagePreview();
        } else {
            throw new Error(data.message || 'Failed to apply filter');
        }
    } catch (error) {
        console.error('Filter error:', error);
        alert(`Error applying filter: ${error.message}`);
        showImageLoading(false);
    }
}

/**
 * Reset all edits to original state
 */
async function resetImage() {
    if (!selectedImage) {
        return;
    }
    
    try {
        showImageLoading(true);
        
        const response = await fetch('/api/editor/image/reset', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                session_id: sessionId
            })
        });
        
        if (!response.ok) {
            throw new Error(`Error resetting image: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            // Update preview and reset controls
            updateImagePreview();
            resetControls();
        } else {
            throw new Error(data.message || 'Failed to reset image');
        }
    } catch (error) {
        console.error('Reset error:', error);
        alert(`Error resetting image: ${error.message}`);
        showImageLoading(false);
    }
}

/**
 * Undo the last edit
 */
async function undoLastEdit() {
    if (!selectedImage) {
        return;
    }
    
    try {
        showImageLoading(true);
        
        const response = await fetch('/api/editor/image/undo', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                session_id: sessionId
            })
        });
        
        if (!response.ok) {
            throw new Error(`Error undoing edit: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            // Update preview
            updateImagePreview();
        } else {
            throw new Error(data.message || 'Nothing to undo');
        }
    } catch (error) {
        console.error('Undo error:', error);
        alert(`Error undoing last edit: ${error.message}`);
        showImageLoading(false);
    }
}

/**
 * Save the edited image
 */
async function saveImage() {
    if (!selectedImage) {
        return;
    }
    
    try {
        showImageLoading(true);
        
        // Get save parameters
        const format = formatSelect.value;
        const quality = parseInt(qualitySlider.value);
        const customFilename = outputFilename.value.trim();
        
        const response = await fetch('/api/editor/image/save', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                session_id: sessionId,
                output_path: customFilename || null,
                format: format,
                quality: quality
            })
        });
        
        if (!response.ok) {
            throw new Error(`Error saving image: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            alert(`Image saved successfully to ${data.path}`);
        } else {
            throw new Error(data.message || 'Failed to save image');
        }
    } catch (error) {
        console.error('Save error:', error);
        alert(`Error saving image: ${error.message}`);
    } finally {
        showImageLoading(false);
    }
}

/**
 * Fetch and display image metadata
 */
async function showMetadata() {
    if (!selectedImage) {
        return;
    }
    
    try {
        metadataContent.innerHTML = '<p class="text-center text-gray-500">Loading metadata...</p>';
        metadataModal.classList.remove('hidden');
        
        // Fetch detailed metadata
        const response = await fetch(`/api/editor/preview/metadata?path=${encodeURIComponent(selectedImage)}`);
        
        if (!response.ok) {
            throw new Error(`Error fetching metadata: ${response.statusText}`);
        }
        
        const data = await response.json();
        
        if (data.success) {
            // Format metadata as HTML
            const meta = data.metadata;
            let html = '<div class="grid grid-cols-2 gap-2">';
            
            // Basic info
            html += `
                <div class="font-medium">Filename:</div><div>${meta.filename}</div>
                <div class="font-medium">Format:</div><div>${meta.format || 'Unknown'}</div>
                <div class="font-medium">Dimensions:</div><div>${meta.size.width} × ${meta.size.height} pixels</div>
                <div class="font-medium">File Size:</div><div>${formatFileSize(meta.file_size)}</div>
                <div class="font-medium">Last Modified:</div><div>${formatDate(meta.modified)}</div>
            `;
            
            // EXIF data if available
            if (meta.exif && Object.keys(meta.exif).length > 0) {
                html += '<div class="col-span-2 mt-4 pt-4 border-t border-gray-200 font-bold">EXIF Metadata</div>';
                
                for (const [key, value] of Object.entries(meta.exif)) {
                    html += `<div class="font-medium">${key}:</div><div>${value}</div>`;
                }
            }
            
            html += '</div>';
            metadataContent.innerHTML = html;
        } else {
            throw new Error(data.message || 'Failed to fetch metadata');
        }
    } catch (error) {
        console.error('Metadata error:', error);
        metadataContent.innerHTML = `<p class="text-red-500">Error fetching metadata: ${error.message}</p>`;
    }
}

// --- UI Functions ---

/**
 * Update the browser UI with directory contents
 */
function updateBrowserUI(contents) {
    // Update current path display
    currentPath.textContent = contents.path;
    
    // Enable/disable parent directory button
    btnParentDir.disabled = !contents.parent;
    
    // Update directory list
    const dirList = document.createElement('ul');
    dirList.className = 'list-none text-sm';
    
    if (contents.directories.length === 0) {
        const li = document.createElement('li');
        li.className = 'text-gray-500';
        li.textContent = 'No subdirectories';
        dirList.appendChild(li);
    } else {
        contents.directories.forEach(dir => {
            const li = document.createElement('li');
            li.className = 'hover:bg-blue-100 cursor-pointer p-1 rounded flex items-center';
            li.innerHTML = `<span class="text-yellow-500 mr-1">📁</span> ${dir.name}`;
            li.addEventListener('click', () => browseDirectory(dir.path));
            dirList.appendChild(li);
        });
    }
    
    directoryList.innerHTML = '';
    directoryList.appendChild(dirList);
    
    // Update image list with thumbnails
    updateImageList(contents.files);
}

/**
 * Update the image list with thumbnails
 */
function updateImageList(files) {
    const container = document.createElement('div');
    container.className = 'grid grid-cols-3 gap-2 image-thumbnails';
    
    if (files.length === 0) {
        const noFiles = document.createElement('div');
        noFiles.className = 'text-gray-500 col-span-3 text-center py-4';
        noFiles.textContent = 'No images found';
        container.appendChild(noFiles);
    } else {
        files.forEach(file => {
            const thumbDiv = document.createElement('div');
            thumbDiv.className = 'thumbnail-container border rounded p-1 cursor-pointer hover:border-blue-500';
            thumbDiv.dataset.path = file.path;
            
            // Add selected class if this is the currently selected image
            if (file.path === selectedImage) {
                thumbDiv.classList.add('border-blue-500', 'bg-blue-50');
            }
            
            // Create thumbnail image
            const img = document.createElement('img');
            img.src = file.thumbnail_url;
            img.alt = file.name;
            img.className = 'w-full h-auto aspect-square object-contain bg-gray-100';
            
            // Add filename below the thumbnail
            const fileName = document.createElement('div');
            fileName.className = 'text-xs truncate mt-1 text-center';
            fileName.title = file.name;
            fileName.textContent = file.name;
            
            // Add click event to load the image
            thumbDiv.addEventListener('click', () => loadImage(file.path));
            
            thumbDiv.appendChild(img);
            thumbDiv.appendChild(fileName);
            container.appendChild(thumbDiv);
        });
    }
    
    imageList.innerHTML = '';
    imageList.appendChild(container);
}

/**
 * Update image information display
 */
function updateImageInfo() {
    if (!selectedImage || !imageMetadata) {
        imageName.textContent = 'No Image Selected';
        return;
    }
    
    // Update image name display
    imageName.textContent = imageMetadata.filename;
    
    // If editing a RAW file, set default output format to JPEG
    if (imageMetadata.is_raw && formatSelect) {
        formatSelect.value = 'JPEG';
    }
    
    // Set default output filename
    const baseName = imageMetadata.filename.substring(0, imageMetadata.filename.lastIndexOf('.'));
    outputFilename.value = `${baseName}_edited`;
}

/**
 * Toggle crop mode
 */
function toggleCropMode(enable) {
    isCropMode = enable;
    
    if (enable) {
        // Show crop controls
        cropControls.classList.remove('hidden');
        
        // Set initial crop values to full image size
        if (originalImageSize) {
            cropLeft.value = 0;
            cropTop.value = 0;
            cropRight.value = originalImageSize[0];
            cropBottom.value = originalImageSize[1];
        }
        
        // Change button state
        btnToggleCrop.textContent = 'Cancel Crop';
    } else {
        // Hide crop controls
        cropControls.classList.add('hidden');
        
        // Change button state
        btnToggleCrop.textContent = 'Crop';
        
        // Update preview to remove any crop preview
        if (cropPreviewMode) {
            cropPreviewMode = false;
            updateImagePreview();
        }
    }
}

/**
 * Enable or disable all controls
 */
function enableControls(enable) {
    // Action buttons
    btnUndo.disabled = !enable;
    btnReset.disabled = !enable;
    btnToggleCrop.disabled = !enable;
    btnRotate.disabled = !enable;
    btnSave.disabled = !enable;
    
    // Adjustment controls
    brightnessSlider.disabled = !enable;
    contrastSlider.disabled = !enable;
    saturationSlider.disabled = !enable;
    filterSelect.disabled = !enable;
    formatSelect.disabled = !enable;
    qualitySlider.disabled = !enable;
    outputFilename.disabled = !enable;
    btnShowMetadata.disabled = !enable;
    
    // Reset crop mode
    if (!enable && isCropMode) {
        toggleCropMode(false);
    }
}

/**
 * Reset all control values to defaults
 */
function resetControls() {
    // Reset adjustment sliders
    brightnessSlider.value = 1.0;
    brightnessValue.textContent = '1.0';
    contrastSlider.value = 1.0;
    contrastValue.textContent = '1.0';
    saturationSlider.value = 1.0;
    saturationValue.textContent = '1.0';
    
    // Reset filter selection
    filterSelect.value = '';
    
    // Exit crop mode if active
    if (isCropMode) {
        toggleCropMode(false);
    }
}

/**
 * Show or hide directory loading state
 */
function showDirectoryLoading(show) {
    // Add loading state to UI elements
    directoryList.classList.toggle('opacity-50', show);
    imageList.classList.toggle('opacity-50', show);
}

/**
 * Show or hide image loading state
 */
function showImageLoading(show) {
    if (show) {
        previewLoading.classList.remove('hidden');
        imagePreview.classList.add('opacity-50');
    } else {
        previewLoading.classList.add('hidden');
        imagePreview.classList.remove('opacity-50');
    }
}

/**
 * Format a file size in bytes to human-readable format
 */
function formatFileSize(bytes) {
    if (bytes < 1024) {
        return bytes + ' bytes';
    } else if (bytes < 1024 * 1024) {
        return (bytes / 1024).toFixed(2) + ' KB';
    } else {
        return (bytes / (1024 * 1024)).toFixed(2) + ' MB';
    }
}

/**
 * Format a timestamp to a readable date
 */
function formatDate(timestamp) {
    return new Date(timestamp * 1000).toLocaleString();
}

// --- Event Listeners ---
document.addEventListener('DOMContentLoaded', () => {
    // Initial page load
    browseDirectory();
    
    // Directory navigation
    btnRefreshImages.addEventListener('click', () => browseDirectory(currentDirectory));
    btnParentDir.addEventListener('click', () => {
        if (btnParentDir.disabled) return;
        browseDirectory(document.getElementById('current-path').dataset.parent);
    });
    
    // Control events
    btnUndo.addEventListener('click', undoLastEdit);
    btnReset.addEventListener('click', resetImage);
    btnToggleCrop.addEventListener('click', () => toggleCropMode(!isCropMode));
    btnSave.addEventListener('click', saveImage);
    
    // Crop controls
    btnCropPreview.addEventListener('click', () => {
        // Implement crop preview logic
        cropPreviewMode = true;
        // This would typically update the preview to show crop outline
        alert('Crop preview not implemented yet');
    });
    btnApplyCrop.addEventListener('click', applyCrop);
    
    // Rotate controls
    btnRotate.addEventListener('click', () => {
        rotationSlider.value = 0;
        rotationValue.textContent = '0°';
        rotateModal.classList.remove('hidden');
    });
    closeRotateModal.addEventListener('click', () => {
        rotateModal.classList.add('hidden');
    });
    btnApplyRotation.addEventListener('click', () => {
        applyRotation(parseInt(rotationSlider.value));
        rotateModal.classList.add('hidden');
    });
    rotationSlider.addEventListener('input', () => {
        rotationValue.textContent = `${rotationSlider.value}°`;
    });
    
    // Adjustment sliders
    brightnessSlider.addEventListener('input', () => {
        brightnessValue.textContent = brightnessSlider.value;
    });
    brightnessSlider.addEventListener('change', applyAdjustments);
    
    contrastSlider.addEventListener('input', () => {
        contrastValue.textContent = contrastSlider.value;
    });
    contrastSlider.addEventListener('change', applyAdjustments);
    
    saturationSlider.addEventListener('input', () => {
        saturationValue.textContent = saturationSlider.value;
    });
    saturationSlider.addEventListener('change', applyAdjustments);
    
    // Filter selection
    filterSelect.addEventListener('change', () => {
        if (filterSelect.value) {
            applyFilter(filterSelect.value);
        }
    });
    
    // Format change
    formatSelect.addEventListener('change', () => {
        // Show/hide quality slider based on format
        const formatValue = formatSelect.value;
        const qualityControl = document.getElementById('quality-control');
        if (formatValue === 'JPEG') {
            qualityControl.classList.remove('hidden');
        } else {
            qualityControl.classList.add('hidden');
        }
    });
    
    // Quality slider
    qualitySlider.addEventListener('input', () => {
        qualityValue.textContent = qualitySlider.value;
    });
    
    // Metadata modal
    btnShowMetadata.addEventListener('click', showMetadata);
    closeMetadataModal.addEventListener('click', () => {
        metadataModal.classList.add('hidden');
    });
    
    // Close modals when clicking outside
    window.addEventListener('click', (event) => {
        if (event.target === metadataModal) {
            metadataModal.classList.add('hidden');
        }
        if (event.target === rotateModal) {
            rotateModal.classList.add('hidden');
        }
    });
});